matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushrules/EventMatchCondition.kt

```diff
@@ -31,40 +29,36 @@ class EventMatchCondition(
          * The glob-style pattern to match against. Patterns with no special glob characters should
          * be treated as having asterisks prepended and appended when testing the condition.
          */
-        val pattern: String,
-        /**
-         * true to match only words. In this case pattern will not be considered as a glob
-         */
-        val wordsOnly: Boolean
+        val pattern: String
 ) : Condition {
 
     override fun isSatisfied(event: Event, conditionResolver: ConditionResolver): Boolean {
         return conditionResolver.resolveEventMatchCondition(event, this)
     }
 
-    override fun technicalDescription() = "'$key' matches '$pattern', words only '$wordsOnly'"
+    override fun technicalDescription() = "'$key' matches '$pattern'"
 
     fun isSatisfied(event: Event): Boolean {
         // TODO encrypted events?
         val rawJson = MoshiProvider.providesMoshi().adapter(Event::class.java).toJsonValue(event) as? Map<*, *>
                 ?: return false
         val value = extractField(rawJson, key) ?: return false
 
-        // Patterns with no special glob characters should be treated as having asterisks prepended
-        // and appended when testing the condition.
+        // The match is performed case-insensitively, and must match the entire value of
+        // the event field given by `key` (though see below regarding `content.body`). The
+        // exact meaning of "case insensitive" is defined by the implementation of the
+        // homeserver.
+        //
+        // As a special case, if `key` is `content.body`, then `pattern` must instead
+        // match any substring of the value of the property which starts and ends at a
+        // word boundary.
         return try {
-            if (wordsOnly) {
-                value.caseInsensitiveFind(pattern)
-            } else {
-                val modPattern = if (pattern.hasSpecialGlobChar()) {
-                    // Regex.containsMatchIn() is way faster without leading and trailing
-                    // stars, that don't make any difference for the evaluation result
-                    pattern.removePrefix("*").removeSuffix("*").simpleGlobToRegExp()
-                } else {
-                    pattern.simpleGlobToRegExp()
-                }
-                val regex = Regex(modPattern, RegexOption.DOT_MATCHES_ALL)
+            if (key == "content.body") {
+                val regex = Regex("(\\W|^)" + pattern.simpleGlobToRegExp() + "(\\W|$)", setOf(RegexOption.DOT_MATCHES_ALL, RegexOption.IGNORE_CASE))
                 regex.containsMatchIn(value)
+            } else {
+                val regex = Regex(pattern.simpleGlobToRegExp(), setOf(RegexOption.DOT_MATCHES_ALL, RegexOption.IGNORE_CASE))
```

Comment:
Shouldn't you keep the existing optimisation to remove * prefix/suffix? as per https://github.com/vector-im/element-android/pull/5008

Reply:
So in theory we could, but in this branch we are not matching the content.body field, but other fields of the event, which usually don't contain a long text. If someone explicitly writes a rule to match `*a*` on the sender field, we could in theory match 255 characters. It would also need to be extended to not do a containsMatchIn, since if there was no `*` at the start of the string, we need to match the start of the string.

The important part here is in the following line, which does `matches` instead of `containsMatchIn`. This would need to change to:

```
if (pattern.startsWith("*") && pattern.endsWith("*")) {
    val regex = Regex(pattern.removePrefix("*").removeSuffix("*").simpleGlobToRegExp(), RegexOption.DOT_MATCHES_ALL)
    regex.containsMatchIn(value)
} else if (pattern.startsWith("*")) {
  // match if value endsWith pattern
} else if (pattern.endsWith("*")) {
  // match if value startsWith pattern
} else {
  // match if complete value matches pattern
}
```

The last 3 branches could probably be collapsed just into one, but considering that fields apart from `content.body` are usually short, patterns usually never start AND end with a `*` and this would mostly affect the user themselves if they find a way to write such a rule that actually is slow to evaluate, I would say it is not worth it in this case.

Now on the other hand in the half that matches on `content.body`, we need to check for word boundaries at the start and end for most matches. Now I don't think this matters if we have a glob on both sides, so I think keeping the optimization can actually be useful there. But for context, the old optimization was because Element Android did `*@room*` to match the body, which this PR explicitly removes, since that matches the wrong stuff. So this only optimizes some badly written user rules and tbh probably should be implemented upstream in the regex engine... Anyway, I'll add the optimization back to the content.body branch.

Reply:
Added the optimization back in the body branch, not in the other branch since I don't really see a need for it, but if you think I should, I can do that as well.